src/wutil/gettext.rs

```diff
@@ -54,114 +256,226 @@ mod gettext_impl {
+///    used. This can only specify a single language, e.g. `LANG=de_AT.UTF-8`.
+///    There, we normalize locale names by stripping off the suffix, leaving only the `ll_CC` part.
+/// 5. Otherwise, localization will not happen.
+///
+/// If users specify `ll_CC` as a language and we don't have a catalog for this language, but we
+/// have one for `ll`, that will be used instead. If users specify `ll` (without specifying a
+/// language variant), which we discourage, and we don't have a catalog for `ll`, but we do have
+/// one for `ll_CC`, that will be used as a fallback. If we have multiple `ll_*` catalogs, all of
+/// them will be used, in arbitrary order.
+#[cfg(feature = "localize-messages")]
+pub fn update_from_env(vars: &EnvStack) {
+    gettext_impl::update_from_env(get_message_locale(vars), get_language_var(vars));
+}
+
+#[cfg(feature = "localize-messages")]
+pub struct SetLocaleLints<'a> {
+    malformed: Vec<&'a str>,
+    duplicates: Vec<&'a str>,
+    non_existing: Vec<&'a str>,
+}
+
+#[cfg(feature = "localize-messages")]
+impl<'a> From<gettext_impl::SetLocaleLints<'a>> for SetLocaleLints<'a> {
+    fn from(lints: gettext_impl::SetLocaleLints<'a>) -> Self {
+        Self {
+            malformed: lints.malformed,
+            duplicates: lints.duplicates,
+            non_existing: lints.non_existing,
+        }
+    }
+}
+
+#[cfg(feature = "localize-messages")]
+impl<'a> SetLocaleLints<'a> {
+    pub fn display_malformed(&self) -> WString {
+        if self.malformed.is_empty() {
+            return WString::new();
+        }
+        wgettext_fmt!(
+            "Language specifiers are not well-formed: %s\n",
+            format!("{:?}", self.malformed)
+        )
+    }
+
+    pub fn display_duplicates(&self) -> WString {
+        if self.duplicates.is_empty() {
+            return WString::new();
+        }
+        wgettext_fmt!(
+            "Language specifiers appear repeatedly: %s\n",
+            format!("{:?}", self.duplicates)
+        )
+    }
+
+    pub fn display_non_existing(&self) -> WString {
+        if self.non_existing.is_empty() {
+            return WString::new();
+        }
+        wgettext_fmt!(
+            "Language specifiers do not have catalogs: %s\n",
+            format!("{:?}", self.non_existing)
+        )
+    }
+
+    pub fn display_all(&self) -> WString {
+        let mut result = WString::new();
+        result.push_utfstr(&self.display_malformed());
+        result.push_utfstr(&self.display_duplicates());
+        result.push_utfstr(&self.display_non_existing());
+        result
+    }
+}
+/// Call this when the `status locale` builtin should update the language precedence.
+/// `langs` should be the list of languages the precedence should be set to.
+#[cfg(feature = "localize-messages")]
+pub fn update_from_status_locale_builtin<'a>(langs: &[&'a str]) -> SetLocaleLints<'a> {
+    gettext_impl::update_from_status_locale_builtin(langs).into()
```

Comment:
ok so the reason for this extra function call is probably to prepare for the fluent backend